test/vidom/tree-test.js

```diff
@@ -0,0 +1,388 @@
+      { block: 'b1' }
+     ],
+    [ 'div', null,
+      [ 'div', { className: 'b1' }, 'first' ],
+      [ 'div', { className: 'b1' }, 'second' ] ],
+    { count: 5 });
+  });
+
+  it('should support early `.once()`', function() {
+    test(function() {
+      block('b1').content()(function() {
+        return 'second';
+      });
+      block('b1').once().match(function() {
+        return !this.ctx.early;
+      }).content()(function() {
+        return 'first';
+      });
+    },
+    [
+      { block: 'b1', early: true },
+      { block: 'b1' }
+     ],
+    [ 'div', null,
+      [ 'div', { className: 'b1' }, 'second' ],
+      [ 'div', { className: 'b1' }, 'first' ] ],
+    { count: 5 });
+  });
+
+  it('should support `.wrap()`', function() {
+    test(function() {
+      block('b1').wrap()(function() {
+        return {
+          block: 'wrap',
+          content: this.ctx
+        }
+      });
+    },
+    [
+      {
+        block: 'b1',
+        tag: 'a',
+        content: {
+          block: 'b1',
+          tag: 'a'
+        }
+      }
+     ],
+    [ 'div', { className: 'wrap' },
+      [ 'a', { className: 'b1' },
+         [ 'div', { className: 'wrap' },
+           [ 'a', { className: 'b1' } ]
+          ]
+       ]
+     ]
+    );
+  });
+
+  it('should support predicates after `wrap()`', function() {
+    test(function() {
+      block('b1')(
+        wrap().match(function() {
+          return this.ctx.key;
+        })(function() {
+          return {
+            block: 'wrap',
+            content: this.ctx
+          }
+        })
+      );
+    },
+    [
+      {
+        block: 'b1',
+        tag: 'a',
+        key: 'val'
+      }
+     ],
+    [ 'div', { className: 'wrap' },
+      [ 'a', { className: 'b1' } ]
+     ]);
+  });
+
+  it('should support `.xjstOptions()`', function() {
+    test(function() {
+      block('b1').xjstOptions({ who: 'cares' }).content()(function() {
+        return 'ok';
+      });
+    },
+    { block: 'b1' },
+    [ 'div', { className: 'b1' }, 'ok' ]);
+  });
+
+  it('should verify match() argument', function() {
+    fail(function() {
+      block('b1').match('123')('123');
+    }, /Wrong.*match.*argument/);
+  });
+
+  it('should execute matches in right order', function() {
+    test(function() {
+      block('bla')(
+        tag()('span'),
+        // this.ctx.d is undefined
+        match(function() { return this.ctx.d; })(
+          tag()('a'),
+          attrs()(match(function() { return this.ctx.d.a; })(function() {
+            // this will throw error
+            return { f: 1 };
+          }))
+        )
+      );
+    },
+    { block: 'bla' },
+    [ 'span', { className: 'bla' } ]);
+  });
+
+  it('should transform array of strings to string', function() {
+    test(function() {
+      block('a')(
+        content()(function() {
+          return [ 'hi', ' there' ];
+        })
+      );
+    },
+    { block: 'a' },
+    [ 'div', { className: 'a' }, 'hi there' ]);
+  });
+});
```

Comment:
We don’t need to have all of this copy-paste tests.

I think, we need write tests for `lib/vidom/index.js` only.